Feature cli (#56)

pxp928 opened the pull request:
Added CLI support for in-toto-golang. Updated Readme and added Makefile support. Supports ITE-7 implementation for certificate constraints.  Functionality testing of the CLI done via the Makefile. Second PR of the Boxboat fork.

*Please fill in the fields below to submit a pull request.  The more
information that is provided, the better.*

**Fixes issue #**: #106


**Description of pull request**:


**Please verify and check that the pull request fulfills the following
requirements**:

- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

shibumi reviewed (approved):
lgtm!

adityasaky reviewed (changes requested):
Thanks for the PR @pxp928! This looks great, and I think we should borrow some things from here for the CLI in https://github.com/in-toto/in-toto. However, I've marked some changes I think we need for consistency etc...

mikhailswift commented:
I think the `key layout` command should be modified to exclude the `private` field of the keyval struct to be consistent with what the layout model expects

adityasaky commented:
Good point @mikhailswift.

pxp928 commented:
I will work through the fixes and changes today. Thanks for the feedback!

pxp928 commented on README.md:
```diff
@@ -53,6 +53,177 @@ func main() {
+Available Commands:
+  help        Help about any command
+  record      Creates a signed link metadata file in two steps, in order to provide
+evidence for supply chain steps that cannot be carried out by a single command
```
Updated!

pxp928 commented on cmd/in-toto/run.go:
```diff
@@ -0,0 +1,149 @@
+}
+
+func run(cmd *cobra.Command, args []string) error {
+	block, err := intoto.InTotoRun(stepName, runDir, materialsPaths, productsPaths, args, key, []string{"sha256"}, []string{}, lStripPaths)
```
Added --exclude to match with python version

pxp928 commented on cmd/in-toto/run.go:
```diff
@@ -0,0 +1,149 @@
+
+	runCmd.Flags().StringVarP(
+		&outDir,
+		"output-directory",
```
changed.

pxp928 commented on cmd/in-toto/run.go:
```diff
@@ -0,0 +1,149 @@
+func runPreRun(cmd *cobra.Command, args []string) error {
+	key = intoto.Key{}
+	cert = intoto.Key{}
+	if err := key.LoadKeyDefaults(keyPath); err != nil {
```
added check and error for file not found.

pxp928 commented on cmd/in-toto/record.go:
```diff
@@ -0,0 +1,161 @@
+	)
+}
+
+func recordPreRun(cmd *cobra.Command, args []string) error {
```
added check and error for file not found.

adityasaky replied:
Same as with run. It should error out if neither is supplied I think? In the python implementation, we require users to supply either a key or to use GPG.

adityasaky replied:
I apologize if my original review was unclear on this front. :smile:

pxp928 replied:
added check if neither key or cert are provided.

pxp928 commented on cmd/in-toto/record.go:
```diff
@@ -0,0 +1,161 @@
+		return fmt.Errorf("failed to create stop link file: %w", err)
+	}
+
+	linkName := fmt.Sprintf(intoto.LinkNameFormat, recordStepName, key.KeyID)
```
added `outdir` for record.

pxp928 commented on cmd/in-toto/record.go:
```diff
@@ -0,0 +1,161 @@
+		return fmt.Errorf("failed to load start link file at %s: %w", prelimLinkName, err)
+	}
+
+	linkMb, err := intoto.InTotoRecordStop(prelimLinkMb, recordProductsPaths, key, []string{"sha256"}, []string{}, nil)
```
passed `--exclude` and `--lstrip-paths` into function.

pxp928 commented:
> I think the `key layout` command should be modified to exclude the `private` field of the keyval struct to be consistent with what the layout model expects

Should this just be changing the `Private` filed to `Private     string `json:"private,omitempty"``? Or did you have some other thoughts in mind?

mikhailswift commented:
> > I think the `key layout` command should be modified to exclude the `private` field of the keyval struct to be consistent with what the layout model expects
> 
> @mikhailswift Should this just be changing the `Private` filed to string `json:"private,omitempty"`? Or did you have some other thoughts in mind?

I think that should do it :)

pxp928 commented:
> > > I think the `key layout` command should be modified to exclude the `private` field of the keyval struct to be consistent with what the layout model expects
> > 
> > 
> > @mikhailswift Should this just be changing the `Private` filed to string `json:"private,omitempty"`? Or did you have some other thoughts in mind?
> 
> I think that should do it :)

@mikhailswift I just did a replace to remove the empty `private` field. Using the omitempty caused many of the tests to fail. Is that what you were looking to do? The `key layout` is now properly formatted with the `private` field.

adityasaky commented:
If `in-toto key layout` is pointed to a private key, it should still only print the public key. As it stands, it currently prints both, because `private` is not empty. The command is meant to be used with the layout, where we only ever add public keys, so we don't want it to ever print the private key.

Also, I wonder if we can enhance it to output `<KEYID>: <KEYOBJ>` rather than just `<KEYOBJ>`. This makes it easier to paste into the pubkeys field of a layout.

adityasaky commented:
I'll take another look at the PR as a whole today afternoon, thanks for addressing the comments @pxp928. :smile:

adityasaky commented on cmd/in-toto/run.go:
```diff
@@ -0,0 +1,168 @@
+func runPreRun(cmd *cobra.Command, args []string) error {
+	key = intoto.Key{}
+	cert = intoto.Key{}
+	if len(keyPath) > 0 {
```
Can we make it so that at least one of key and cert needs to be provided?

pxp928 replied:
added check to make sure a key or cert is provided.

pxp928 replied:
added check if neither key or cert are provided.

adityasaky commented on README.md:
````diff
@@ -53,6 +53,212 @@ func main() {
+Use "in-toto key [command] --help" for more information about a command.
+```
+
+### run
````
Exclude should be added. Also, needs to reflect `--output-directory` -> `--metadata-directory` change. Applies to record as well.

pxp928 replied:
Changed to `--metadata-directory` for both run and record. `--exclude` is already added for both run and record.

pxp928 commented:
> If `in-toto key layout` is pointed to a private key, it should still only print the public key. As it stands, it currently prints both, because `private` is not empty in this case. The command is meant to be used with the layout, where we only ever add public keys, so we don't want it to ever print the private key.
> 
> Also, I wonder if we can enhance it to output `<KEYID>: <KEYOBJ>` rather than just `<KEYOBJ>`. This makes it easier to paste into the pubkeys field of a layout.

Updated such that `in-toto key layout` now does not print out the `private` field. Also changed the output to `<KEYID>: <KEYOBJ>` such that it can be added directly to the layout. Changed the Makefile to reflect this new substitution into the layout for testing.

adityasaky reviewed (approved):
Thanks @pxp928! This looks fine to me. I'm happy to merge tonight or so, but if you get a chance before then, could I ask you to squash some commits?

adityasaky commented:
Thanks for your work on this, @pxp928!
